src/lib/operators/join_sort_merge.cpp

```diff
@@ -845,14 +1160,72 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractReadOnlyOperatorImpl {
     }
   }
 
+  bool _check_if_table_contains_null(const Table& table) {
+    const auto& table_column_definition =
+        table.column_definitions()[_sort_merge_join._primary_predicate.column_ids.second];
+    if (!table_column_definition.nullable) return false;
+
+    const auto chunk_count = table.chunk_count();
+    auto has_null = false;
+
+    for (auto cluster_id = ChunkID{0}; cluster_id < chunk_count; ++cluster_id) {
+      const auto& chunk = table.get_chunk(cluster_id);
+      const auto& segment = chunk->get_segment(_sort_merge_join._primary_predicate.column_ids.second);
+      segment_iterate<T>(*segment, [&](const auto& position) {
+        if (position.is_null()) {
+          has_null = true;
+          return;
+        }
+      });
+      if (has_null) return true;
+    }
+    return false;
+  }
+
  public:
   std::shared_ptr<const Table> _on_execute() override {
-    const auto include_null_left = (_mode == JoinMode::Left || _mode == JoinMode::FullOuter);
-    const auto include_null_right = (_mode == JoinMode::Right || _mode == JoinMode::FullOuter);
-    auto radix_clusterer = RadixClusterSort<T>(
-        _sort_merge_join.left_input_table(), _sort_merge_join.right_input_table(),
-        _sort_merge_join._primary_predicate.column_ids, _primary_predicate_condition == PredicateCondition::Equals,
-        include_null_left, include_null_right, _cluster_count, _performance);
+    if (_mode == JoinMode::Semi) {
+      if (_primary_predicate_condition == PredicateCondition::NotEquals &&
+          _sort_merge_join.right_input_table()->empty()) {
+        return _sort_merge_join.left_input_table();
+      }
+      if (_primary_predicate_condition == PredicateCondition::Equals && _sort_merge_join.right_input_table()->empty()) {
+        return Table::create_dummy_table(_sort_merge_join.left_input_table()->column_definitions());
+      }
+    }
+    if (_mode == JoinMode::AntiNullAsFalse && _sort_merge_join.right_input_table()->empty()) {
+      return _sort_merge_join.left_input_table();
+    }
+    if (_mode == JoinMode::AntiNullAsTrue) {
+      if (_sort_merge_join.right_input_table()->empty()) {
+        return _sort_merge_join.left_input_table();
+      }
+      if (_secondary_join_predicates.empty()) {
+        auto right_side_has_null = _check_if_table_contains_null(*(_sort_merge_join.right_input_table()));
+        if (right_side_has_null) {
+          return Table::create_dummy_table(_sort_merge_join.left_input_table()->column_definitions());
+        }
+      }
+    }
+
+    auto sort = true;
+    if (_mode == JoinMode::AntiNullAsFalse || _mode == JoinMode::AntiNullAsTrue) {
+      // Since the not equal anti join does not support multiple predicates we are only interested in equal values. For
+      // that, we do not need to sort the tables.
+      sort = !(_primary_predicate_condition == PredicateCondition::NotEquals);
```

Comment:
Similar to the semi joins: I would remove the non-equal cases for simplicity and add a comment to the supports function, that this PR contains code for it but we discarded it for simplicity.